pkg/publicshare/manager/cs3/cs3.go

```diff
@@ -0,0 +1,401 @@
+// Manager implements a publicshare manager using a cs3 storage backend
+type Manager struct {
+	storage          metadata.Storage
+	indexer          indexer.Indexer
+	passwordHashCost int
+
+	initialized bool
+}
+
+// PublicShareWithPassword represents a public share including its hashes password
+type PublicShareWithPassword struct {
+	PublicShare    *link.PublicShare `json:"public_share"`
+	HashedPassword string            `json:"password"`
+}
+
+type config struct {
+	GatewayAddr       string `mapstructure:"gateway_addr"`
+	ProviderAddr      string `mapstructure:"provider_addr"`
+	ServiceUserID     string `mapstructure:"service_user_id"`
+	ServiceUserIdp    string `mapstructure:"service_user_idp"`
+	MachineAuthAPIKey string `mapstructure:"machine_auth_apikey"`
+}
+
+// NewDefault returns a new manager instance with default dependencies
+func NewDefault(m map[string]interface{}) (publicshare.Manager, error) {
+	c := &config{}
+	if err := mapstructure.Decode(m, c); err != nil {
+		err = errors.Wrap(err, "error creating a new manager")
+		return nil, err
+	}
+
+	s, err := metadata.NewCS3Storage(c.GatewayAddr, c.ProviderAddr, c.ServiceUserID, c.ServiceUserIdp, c.MachineAuthAPIKey)
+	if err != nil {
+		return nil, err
+	}
+	indexer := indexer.CreateIndexer(s)
+
+	return New(s, indexer, bcrypt.DefaultCost)
+}
+
+// New returns a new manager instance
+func New(storage metadata.Storage, indexer indexer.Indexer, passwordHashCost int) (publicshare.Manager, error) {
+	return &Manager{
+		storage:          storage,
+		indexer:          indexer,
+		passwordHashCost: passwordHashCost,
+		initialized:      false,
+	}, nil
+}
+
+func (m *Manager) initialize() error {
+	if m.initialized {
+		return nil
+	}
+
+	err := m.storage.Init(context.Background(), "public-share-manager-metadata")
+	if err != nil {
+		return err
+	}
+	if err := m.storage.MakeDirIfNotExist(context.Background(), "publicshares"); err != nil {
+		return err
+	}
+	err = m.indexer.AddIndex(&link.PublicShare{}, option.IndexByField("Id.OpaqueId"), "Token", "publicshares", "unique", nil, true)
+	if err != nil {
+		return err
+	}
+	err = m.indexer.AddIndex(&link.PublicShare{}, option.IndexByFunc{
+		Name: "Owner",
+		Func: indexOwnerFunc,
+	}, "Token", "publicshares", "non_unique", nil, true)
+	if err != nil {
+		return err
+	}
+	m.initialized = true
+	return nil
+}
+
+// CreatePublicShare creates a new public share
+func (m *Manager) CreatePublicShare(ctx context.Context, u *user.User, ri *provider.ResourceInfo, g *link.Grant) (*link.PublicShare, error) {
+	if err := m.initialize(); err != nil {
+		return nil, err
+	}
+
+	id := &link.PublicShareId{
+		OpaqueId: utils.RandString(15),
+	}
+
+	tkn := utils.RandString(15)
+	now := time.Now().UnixNano()
+
+	displayName := tkn
+	if ri.ArbitraryMetadata != nil {
+		metadataName, ok := ri.ArbitraryMetadata.Metadata["name"]
+		if !ok {
+			displayName = metadataName
+		}
+	}
+
+	var passwordProtected bool
+	password := g.Password
+	if len(password) > 0 {
```

Comment:
Use `password != ""` instead `len(password) > 0`

It makes it clearer that we are working with a string in this case